### PR TITLE
Contracts Refactor Part 1

### PR DESCRIPTION
**Pull request by zakhap**
## Description

In a moment of clarity, I realized this refactor to a multi-contract system could be done in two parts:
1. the backend can be prepped and migrated to handle multiple contracts with multiple chain-nodes for each community
2. the frontend controller chain-contract-initialization system can be prepped to handle more than one contract at a time and to toggle between them gracefully etc.
This PR only does the first step.
- Creates "Contracts" and "CommunityContracts" tables. The first for each individual contract, the second as a through-table mapping of communities=>contracts. 
- Migration that handles moving of `address` and needed contract information to the "Contracts" tables, creates single row in "CommunityContracts" for each. 
- Updates `/status` route to return array (currently only of one) of `Contract` models. 
- Updates client's `ChainInfo.fromJSON()` method to parse `Contract[]=>address` (maintaining current expectations within the controller system.
- Updates `/createChain` route to findOrCreate a Contract and create a CommunityContract row instead of adding the Contract address directly to the Chain. 
- Updates `validateThreshold` checks to find and pass in Contract now that `chain.address` does not exist.
- Updates `/getTokenCommunity` route to handle new system accordingly.

QA Needed:
- Creating Token communities from UI Form: 
  - [ ] from New Token Contract (not currently in DB)
  - [ ] from Token Contract in DB (creating a new community that _shares a contract with another community_.
 - Make sure dydx & aave & eth governance communities still work.
 
## Motivation and Context

Multi-chain, Multi-Token, Multi-Contract future is coming.

## How has this been tested?

Briefly checked to make sure in compiles, typechecks, and that eth governance pages still load. Needs more thorough QA though. 

## Does this PR affect any server routes?
- [x] yes
- [ ] no

## Have proper tags been added (for bug, enhancement, breaking change)?
- [x] yes

### CODE REVIEW

**Comment by jnaviask** [comment]
Before I take a look: any additional work needed before merge? Or is this ready to go as-is?

**Comment by zakhap** [comment]
> Before I take a look: any additional work needed before merge? Or is this ready to go as-is (assuming it passes QA)?

It should work as is as long as no community maliciously adds a second contract somehow, then only one will work (whichever is returned as the first array item). 
TLDR: No additional work, just check t crossings and i dottings.

File: packages/commonwealth/client/scripts/app.ts
Comment (jnaviask) [inline packages/commonwealth/client/scripts/app.ts:400]: Remove console.log.

File: packages/commonwealth/server/migrations/20220811142850-add-contracts-tables.js
Comment (jnaviask) [inline packages/commonwealth/server/migrations/20220811142850-add-contracts-tables.js:47]: TODO **

File: packages/commonwealth/server/migrations/20220811142850-add-contracts-tables.js
Comment (jnaviask) [inline packages/commonwealth/server/migrations/20220811142850-add-contracts-tables.js:65]: When do we expect to see this?

File: packages/commonwealth/server/migrations/20220811142850-add-contracts-tables.js
Comment (jnaviask) [inline packages/commonwealth/server/migrations/20220811142850-add-contracts-tables.js:122]: Did these have any data prior?

File: packages/commonwealth/server/migrations/20220811142850-add-contracts-tables.js
Comment (jnaviask) [inline packages/commonwealth/server/migrations/20220811142850-add-contracts-tables.js:92]: TODO**

File: packages/commonwealth/server/models/chain.ts
Comment (jnaviask) [inline packages/commonwealth/server/models/chain.ts:165]: No need for this hasMany since we're using it as a through-table in the above.

File: packages/commonwealth/server/models/chain.ts
Comment (jnaviask) [inline packages/commonwealth/server/models/chain.ts:81]: Yep.

File: packages/commonwealth/server/models/chain_node.ts
Comment (jnaviask) [inline packages/commonwealth/server/models/chain_node.ts:15]: Ideally we use a string enum here -- but this is fine for now.

File: packages/commonwealth/server/models/community_contract.ts
Comment (jnaviask) [inline packages/commonwealth/server/models/community_contract.ts:14]: should add `Chain` here as well.

File: packages/commonwealth/server/models/contract.ts
Comment (jnaviask) [inline packages/commonwealth/server/models/contract.ts:15]: What are the possible types?

File: packages/commonwealth/server/models/contract.ts
Comment (jnaviask) [inline packages/commonwealth/server/models/contract.ts:57]: Ideally this should be the same but inverse of the belongsToMany in Chains, via the through table. See: https://sequelize.org/docs/v6/advanced-association-concepts/advanced-many-to-many/

File: packages/commonwealth/server/routes/createChain.ts
Comment (jnaviask) [inline packages/commonwealth/server/routes/createChain.ts:160]: TODO**

File: packages/commonwealth/server/routes/getTokenForum.ts
Comment (jnaviask) [inline packages/commonwealth/server/routes/getTokenForum.ts:86]: We may need to toJSON() here.

File: packages/commonwealth/server/routes/status.ts
Comment (jnaviask) [inline packages/commonwealth/server/routes/status.ts:39]: Should we require ChainNode here as well?

File: packages/commonwealth/server/routes/tokenBalance.ts
Comment (jnaviask) [inline packages/commonwealth/server/routes/tokenBalance.ts:71]: Unrelated -- this is for ERC20 vs ERC721, base will not reflect this.

File: packages/commonwealth/server/routes/tokenBalance.ts
Comment (jnaviask) [inline packages/commonwealth/server/routes/tokenBalance.ts:65]: How do we get this chain_node_id?

File: packages/commonwealth/server/scripts/setupChainEventListeners.ts
Comment (jnaviask) [inline packages/commonwealth/server/scripts/setupChainEventListeners.ts:177]: TODO -- not sure what we should do with this file. May be deprecated post-Timothee's changes.

File: packages/commonwealth/server/util/validateTopicThreshold.ts
Comment (jnaviask) [inline packages/commonwealth/server/util/validateTopicThreshold.ts:38]: You should be able to query this directly from the Chain rather than explicitly referencing the through table.

**Review by jnaviask — CHANGES_REQUESTED** [review]
Added TODO placeholders and several implementation comments.

File: packages/commonwealth/server/models/contract.ts
Comment (zakhap) [reply packages/commonwealth/server/models/contract.ts:15]: currently, `types` are just our network from the chain model, what we use to identify the controller initialization steps. there will be a few more in the list once i get the solana tokens moved over but this this the current: 
```
 aave
 erc721
 compound
 marlin-testnet
 erc20
```

File: packages/commonwealth/server/routes/tokenBalance.ts
Comment (zakhap) [reply packages/commonwealth/server/routes/tokenBalance.ts:65]: from the chain returned at the top of the file from `validateChain()`.

File: packages/commonwealth/server/routes/status.ts
Comment (zakhap) [reply packages/commonwealth/server/routes/status.ts:39]: we could!

File: packages/commonwealth/server/models/contract.ts
Comment (jnaviask) [reply packages/commonwealth/server/models/contract.ts:15]: Can we use an enum somewhere to make it clear what the list of valid types are?

File: packages/commonwealth/server/routes/tokenBalance.ts
Comment (jnaviask) [reply packages/commonwealth/server/routes/tokenBalance.ts:65]: What if the contract we're checking uses a different node id?

Comment (zakhap) [reply packages/commonwealth/server/routes/tokenBalance.ts:65]: fixed

File: packages/commonwealth/server/util/validateTopicThreshold.ts
Comment (zakhap) [reply packages/commonwealth/server/util/validateTopicThreshold.ts:38]: gonna leave it now because we're not fetching the ChainInstance, but receiving a ChainAttribute via the Topic's include statement. moreover, this route doesn't matter long-term as it'll get deprecated with the next Rules push.

File: packages/commonwealth/server/scripts/setupChainEventListeners.ts
Comment (zakhap) [reply packages/commonwealth/server/scripts/setupChainEventListeners.ts:177]: I think this is definitely a post-timothee change.

**Comment by zakhap** [comment]
@timolegros how do you think this will mess with your chain-events refactor?

File: packages/commonwealth/server/scripts/setupChainEventListeners.ts
Comment (zakhap) [reply packages/commonwealth/server/scripts/setupChainEventListeners.ts:177]: @timolegros

Comment (timolegros) [reply packages/commonwealth/server/scripts/setupChainEventListeners.ts:177]: This file is deleted in the Chain-Events microservice PR.

**Comment by timolegros** [comment]
> @timolegros how do you think this will impact your chain-events refactor?

The one issue I see is that if addresses are separated from Chains then I need to keep track of contracts in chain-events which means adding a new queue/producer/consumer. This isn't hard to do but it does increase the complexity of the system as a whole i.e. we now have 6 queues and a producer + consumer function for each one. So not a big deal short term but does indicate that we should think long-term about improving our data models such that the services can be more independent. I'd like to one day have contracts (and maybe even chainNodes) be in the chain-events microservice rather than the main service since the contract model stores almost exclusively chain-related data.

**Comment by zakhap** [comment]
> > @timolegros how do you think this will impact your chain-events refactor?
> 
> The one issue I see is that if addresses are separated from Chains then I need to keep track of contracts in chain-events which means adding a new queue/producer/consumer. This isn't hard to do but it does increase the complexity of the system as a whole i.e. we now have 6 queues and a producer + consumer function for each one. So not a big deal short term but does indicate that we should think long-term about improving our data models such that the services can be more independent. I'd like to one day have contracts (and maybe even chainNodes) be in the chain-events microservice rather than the main service since the contract model stores almost exclusively chain-related data.

I think the difference is keeping track of Contracts instead of _Chain with Addresses_. @jnaviask and I have maybe-not-jokingly discussed having all the core community-chain-chainnode-contract-type data in a service/api of its own that chain-events and cw both consume/utilize.

**Comment by zakhap** [comment]
I believe this is ready for another review. Merging and deploying should be scheduled with @timolegros regarding his Chain-Events PR too, making sure that everything is prepped for both to not be hitched on each other confusingly.

File: packages/commonwealth/server/migrations/20220811142850-add-contracts-tables.js
Comment (jnaviask) [inline packages/commonwealth/server/migrations/20220811142850-add-contracts-tables.js:4]: Migration must include indexes as well, e.g. addIndex, or directly at table creation.

File: packages/commonwealth/server/migrations/20220811142850-add-contracts-tables.js
Comment (zakhap) [reply packages/commonwealth/server/migrations/20220811142850-add-contracts-tables.js:48]: fixed

File: packages/commonwealth/server/migrations/20220811142850-add-contracts-tables.js
Comment (zakhap) [reply packages/commonwealth/server/migrations/20220811142850-add-contracts-tables.js:4]: added `address` index.

File: packages/commonwealth/server/routes/tokenBalance.ts
Comment (zakhap) [reply packages/commonwealth/server/routes/tokenBalance.ts]: Just fixed this, finally.

File: packages/commonwealth/server/models/contract_abi.ts
Comment (jnaviask) [inline packages/commonwealth/server/models/contract_abi.ts:8]: Should this be object or string?

File: packages/commonwealth/server/models/contract_abi.ts
Comment (jnaviask) [inline packages/commonwealth/server/models/contract_abi.ts:26]: We may want to specify the source (etherscan vs user-provided).

File: packages/commonwealth/server/routes/createChain.ts
Comment (jnaviask) [inline packages/commonwealth/server/routes/createChain.ts:59]: Address should be optional rather than mandatory arg.

File: packages/commonwealth/server/routes/createChain.ts
Comment (jnaviask) [inline packages/commonwealth/server/routes/createChain.ts:162]: Do we protect against creating duplicate Cosmos/Substrate chains?

File: packages/commonwealth/server/routes/tokenBalance.ts
Comment (jnaviask) [inline packages/commonwealth/server/routes/tokenBalance.ts:51]: We should make this check positive (i.e. only look for contract in certain circumstances) rather than negative (i.e. always look except in certain circumstances).

File: packages/commonwealth/server/scripts/setupChainEventListeners.ts
Comment (jnaviask) [reply packages/commonwealth/server/scripts/setupChainEventListeners.ts:177]: @zakhap is this PR blocked on CE Microservice PR then?

File: packages/commonwealth/server/util/validateTopicThreshold.ts
Comment (jnaviask) [inline packages/commonwealth/server/util/validateTopicThreshold.ts:49]: What happens if this is undefined?

**Review by jnaviask — COMMENTED** [review]
A few, smaller comments then last time. Then a final pass of QA + merge.

File: packages/commonwealth/server/scripts/setupChainEventListeners.ts
Comment (zakhap) [reply packages/commonwealth/server/scripts/setupChainEventListeners.ts:177]: i don't use this file in any meaningful way in my PR, was just updating it to "work right" with current chain events. being that communities only have 1 contract by default, this should have no effect. I can be redundantly clear at fetching the contract that the contract needs to be of a particular type. Will do so

File: packages/commonwealth/server/models/contract_abi.ts
Comment (zakhap) [reply packages/commonwealth/server/models/contract_abi.ts:26]: should we just have a `verified` column then? and we can periodically update + verify against etherscan accordingly?

File: packages/commonwealth/server/routes/createChain.ts
Comment (zakhap) [reply packages/commonwealth/server/routes/createChain.ts:162]: I don't believe so, but this check by using `req.body.address` wasn't doing so either

File: packages/commonwealth/server/util/validateTopicThreshold.ts
Comment (zakhap) [reply packages/commonwealth/server/util/validateTopicThreshold.ts:49]: i assumed `tbc.getBalance()` would throw or something, will look into it

File: packages/commonwealth/server/routes/createChain.ts
Comment (jnaviask) [reply packages/commonwealth/server/routes/createChain.ts:162]: It would be, because it'll look only for duplicate chain nodes if address is empty

File: packages/commonwealth/server/models/contract_abi.ts
Comment (jnaviask) [reply packages/commonwealth/server/models/contract_abi.ts:26]: reasonable solution, unless we anticipate having different verified / user-submitted ABIs

File: packages/commonwealth/client/scripts/models/Contract.ts
Comment (zakhap) [reply packages/commonwealth/client/scripts/models/Contract.ts:15]: 👍

File: packages/commonwealth/server/models/contract_abi.ts
Comment (zakhap) [reply packages/commonwealth/server/models/contract_abi.ts:8]: is `Record<string, unknown>` valid here in your mind?

File: packages/commonwealth/client/scripts/models/ChainInfo.ts
Comment (jnaviask) [inline packages/commonwealth/client/scripts/models/ChainInfo.ts:305]: Why not use the optional address here, if available?

File: packages/commonwealth/server/scripts/dbNode.ts
Comment (jnaviask) [inline packages/commonwealth/server/scripts/dbNode.ts:146]: This type seems very arbitrary?

File: packages/commonwealth/server/scripts/setupChainEventListeners.ts
Comment (jnaviask) [inline packages/commonwealth/server/scripts/setupChainEventListeners.ts:195]: May need to specify a "requires" in the call if you want to guarantee it grabs ChainNode.

**Review by jnaviask — COMMENTED** [review]
A couple comments -- moving onto QA.

**Comment by jnaviask** [comment]
Looks like chain events migration is broken, which is bad because this is the main tool we use to recover from downtime / disaster scenarios involving missed proposals, etc.

```
$ ENTITY_MIGRATION=dydx yarn migrate-event
yarn run v1.22.19
$ ts-node --project tsconfig.node.json server/scripts/migrateChainEntities.ts

/home/jake/cw/commonwealth/packages/commonwealth/node_modules/ts-node/src/index.ts:421
    return new TSError(diagnosticText, diagnosticCodes)
           ^
TSError: ⨯ Unable to compile TypeScript:
server/scripts/migrateChainEntities.ts:70:23 - error TS2551: Property 'address' does not exist on type 'ChainInstance'. Did you mean 'Addresses'?

70         chainInstance.address
                         ~~~~~~~

  server/models/chain.ts:52:3
    52   Addresses?: AddressAttributes[] | AddressAttributes['id'][];
         ~~~~~~~~~
    'Addresses' is declared here.
server/scripts/migrateChainEntities.ts:77:23 - error TS2551: Property 'address' does not exist on type 'ChainInstance'. Did you mean 'Addresses'?

77         chainInstance.address
                         ~~~~~~~

  server/models/chain.ts:52:3
    52   Addresses?: AddressAttributes[] | AddressAttributes['id'][];
         ~~~~~~~~~
    'Addresses' is declared here.

    at createTSError (/home/jake/cw/commonwealth/packages/commonwealth/node_modules/ts-node/src/index.ts:421:12)
    at reportTSError (/home/jake/cw/commonwealth/packages/commonwealth/node_modules/ts-node/src/index.ts:425:19)
    at getOutput (/home/jake/cw/commonwealth/packages/commonwealth/node_modules/ts-node/src/index.ts:553:36)
    at Object.compile (/home/jake/cw/commonwealth/packages/commonwealth/node_modules/ts-node/src/index.ts:758:32)
    at Module.m._compile (/home/jake/cw/commonwealth/packages/commonwealth/node_modules/ts-node/src/index.ts:837:43)
    at Module._extensions..js (internal/modules/cjs/loader.js:1114:10)
    at Object.require.extensions.<computed> [as .ts] (/home/jake/cw/commonwealth/packages/commonwealth/node_modules/ts-node/src/index.ts:840:12)
    at Module.load (internal/modules/cjs/loader.js:950:32)
    at Function.Module._load (internal/modules/cjs/loader.js:790:12)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:76:12)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Comment by jnaviask** [comment]
I get the following trying to comment on Juno:

```
Trace: TypeError: Cannot read property 'id' of undefined
    at Object.tokenBalance (/home/jake/cw/commonwealth/packages/commonwealth/server/routes/tokenBalance.ts:50:39)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at /home/jake/cw/commonwealth/packages/commonwealth/server/scripts/setupErrorHandlers.ts:40:15
    at newFn (/home/jake/cw/commonwealth/node_modules/express-async-errors/index.js:16:20)
    at Layer.handle_error (/home/jake/cw/commonwealth/node_modules/express/lib/router/layer.js:71:5)
    at trim_prefix (/home/jake/cw/commonwealth/node_modules/express/lib/router/index.js:315:13)
    at /home/jake/cw/commonwealth/node_modules/express/lib/router/index.js:284:7
    at Function.process_params (/home/jake/cw/commonwealth/node_modules/express/lib/router/index.js:335:12)
    at next (/home/jake/cw/commonwealth/node_modules/express/lib/router/index.js:275:10)
    at Layer.handle_error (/home/jake/cw/commonwealth/node_modules/express/lib/router/layer.js:67:12)
    at trim_prefix (/home/jake/cw/commonwealth/node_modules/express/lib/router/index.js:315:13)
    at /home/jake/cw/commonwealth/node_modules/express/lib/router/index.js:284:7
    at Function.process_params (/home/jake/cw/commonwealth/node_modules/express/lib/router/index.js:335:12)
    at next (/home/jake/cw/commonwealth/node_modules/express/lib/router/index.js:275:10)
    at cb (/home/jake/cw/commonwealth/node_modules/rollbar/src/server/rollbar.js:264:14)
    at /home/jake/cw/commonwealth/node_modules/rollbar/src/utility.js:404:7
    at Queue.<anonymous> (/home/jake/cw/commonwealth/node_modules/rollbar/src/queue.js:99:7)
    at Queue.<anonymous> (/home/jake/cw/commonwealth/node_modules/rollbar/src/queue.js:161:9)
POST /api/tokenBalance 500 186.263 ms - 50
```

File: packages/commonwealth/server/scripts/dbNode.ts
Comment (zakhap) [reply packages/commonwealth/server/scripts/dbNode.ts:146]: actually, crazy thought: what if we get rid of the where clause altogether? same set of 11 chains get returned.

File: packages/commonwealth/server/scripts/setupChainEventListeners.ts
Comment (zakhap) [reply packages/commonwealth/server/scripts/setupChainEventListeners.ts:195]: we don't use it here, because we defer to the chain's node as `node.ChainNode.url` instead (always fine currently). gonna remove the comment

**Comment by zakhap** [comment]
I believe I fixed the migration script and the questions at hand. No need for the "requires" above, we just use the chainNode from the chain, rather than the chain node from the contract itself.

**Comment by zakhap** [comment]
Fixed the TBC issue in posting in Juno by returning the ChainNode with the Chain in the `ValidateChain` route

File: packages/commonwealth/server/scripts/dbNode.ts
Comment (zakhap) [reply packages/commonwealth/server/scripts/dbNode.ts]: i removed the WHERE clause as I don't think it's necessary

Comment (zakhap) [reply packages/commonwealth/server/scripts/dbNode.ts]: never mind...

**Comment by jnaviask** [comment]
> I believe I fixed the migration script and the questions at hand. No need for the "requires" above, we just use the chainNode from the chain, rather than the chain node from the contract itself.

Now I get (for dydx and aave migration):
```
2022-09-27 16:37:59,069 INFO [storageFetcher.ts::aave] Fetching Aave entities for range: 0-15627117.
2022-09-27 16:38:00,258 ERROR [migrateChainEntities.ts] Failed to fetch events for aave: CONNECTION ERROR: Provider started to reconnect before the response got received!
```

Not sure if this related to your changes or not.

**Comment by jnaviask** [comment]
> Fixed the TBC issue in posting in Juno by returning the ChainNode with the Chain in the `ValidateChain` route

Still getting these:
```
2022-09-27 16:40:48,525 INFO [tokenBalance.ts] Failed to query token balance: Cannot read property 'address' of undefined
POST /api/selectChain 200 7.568 ms - 20
GET /api/activeThreads?chain=juno&threads_per_topic=5 200 28.241 ms - -
Trace: ServerError: Balance query failed
    at Object.tokenBalance (/home/jake/cw/commonwealth/packages/commonwealth/server/routes/tokenBalance.ts:79:11)
    at processTicksAndRejections (internal/process/task_queues.js:95:5) {
  status: 500,
  error: undefined
}
    at /home/jake/cw/commonwealth/packages/commonwealth/server/scripts/setupErrorHandlers.ts:26:15
    at newFn (/home/jake/cw/commonwealth/node_modules/express-async-errors/index.js:16:20)
    at Layer.handle_error (/home/jake/cw/commonwealth/node_modules/express/lib/router/layer.js:71:5)
    at trim_prefix (/home/jake/cw/commonwealth/node_modules/express/lib/router/index.js:315:13)
    at /home/jake/cw/commonwealth/node_modules/express/lib/router/index.js:284:7
    at Function.process_params (/home/jake/cw/commonwealth/node_modules/express/lib/router/index.js:335:12)
    at next (/home/jake/cw/commonwealth/node_modules/express/lib/router/index.js:275:10)
    at Layer.handle_error (/home/jake/cw/commonwealth/node_modules/express/lib/router/layer.js:67:12)
    at trim_prefix (/home/jake/cw/commonwealth/node_modules/express/lib/router/index.js:315:13)
    at /home/jake/cw/commonwealth/node_modules/express/lib/router/index.js:284:7
    at Function.process_params (/home/jake/cw/commonwealth/node_modules/express/lib/router/index.js:335:12)
    at next (/home/jake/cw/commonwealth/node_modules/express/lib/router/index.js:275:10)
    at cb (/home/jake/cw/commonwealth/node_modules/rollbar/src/server/rollbar.js:264:14)
    at /home/jake/cw/commonwealth/node_modules/rollbar/src/utility.js:404:7
    at Queue.<anonymous> (/home/jake/cw/commonwealth/node_modules/rollbar/src/queue.js:99:7)
    at Queue.<anonymous> (/home/jake/cw/commonwealth/node_modules/rollbar/src/queue.js:161:9)
POST /api/tokenBalance 500 199.391 ms - 62
```

**Review by jnaviask — APPROVED** [review]
I think we can go ahead with this whenever. No obvious major issues, although small ones might arise later on.

**Pull request merged**
